R/track_rust_source.R

```diff
@@ -0,0 +1,110 @@
+# Converts any path to path, relative to the package root
+# E.g., outer_root/some_folder/code/packages/my_package/src/rust/src/lib.rs
+# becomes src/rust/src/lib.rs.
+# Used for pretty printing.
+pretty_rel_path <- function(path, search_from = ".") {
+  fs::path_rel(path, start = rprojroot::find_package_root_file(path = search_from))
+}
+
+get_library_path <- function(path = ".") {
+  # Constructs path to the library file (e.g., package_name.dll)
+  pkg <- desc::desc(rprojroot::find_package_root_file("DESCRIPTION", path = path))
+  pkg_name <- pkg$get("Package")
+  fs::path(
+    glue::glue(
+      rprojroot::find_package_root_file("src", path = path),
+      "{pkg_name}{.Platform$dynlib.ext}",
+      .sep = .Platform$file.sep
+    )
+  )
+}
+
+get_rust_files <- function(path = ".") {
+  # Enumerates all Rust files, changes to which should trigger re-compilation.
+
+  src_root <- rprojroot::find_package_root_file("src", path = path)
+  if (!fs::dir_exists(src_root)) {
+    # No source code found
+    cli::cli_alert_warning("{.file src} directory is missing. Are you sure the package is set up to use Rust?")
+    return(fs::path())
+  }
+
+  # Recursively enumerating files wihtin pkg_root/src
+
+  # This handles Cargo.toml files,
+  cargo_toml_paths <- fs::dir_ls(
+    path = src_root,
+    recurse = TRUE,
+    glob = "*Cargo.toml"
+  )
+
+  # This handles *rs files
+  rust_src_paths <- fs::dir_ls(
+    path = src_root,
+    recurse = TRUE,
+    glob = "*.rs"
+  )
+
+  result <- fs::fs_path(c(cargo_toml_paths, rust_src_paths))
+  result
+}
+
+#' Checks if re-compilation is needed.
+#' Tracks changes in Rust source files (`*.rs`) and `Cargo.toml`.
+#'
+#' @param path Package root.
+#' @param quiet Logical scalar indicating wether the output should be quiet (`TRUE`)
+#'   or verbose (`FALSE`).
+#' @returns Logical `TRUE` if Rust source has been modified, `FALSE` otherwise.
+#' @export
+needs_compilation <- function(path = ".", quiet = FALSE) {
```

Comment:
Do we need to export this function? At a minimum, add `@keywords internal` so it gets excluded from the documentation index.

Reply:
That depends. My reasoning was `pkgbuild::needs_compile()` is exported, so perhaps we can provide a similar API for Rust code. The function has no side effects so can be safely used by a package developer. I can easily remove `@export`.